### ISSUE TITLE
add high-level, cross-platform close-on-exit togglers to twisted.internet.fdesc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3576]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3576|
|Type|enhancement|
|Created|2008-12-12 23:12:47Z|
|Branch|https://github.com/twisted/twisted/tree/cloexec-3576|

The internet file descriptors created by reactors shouldn't be inherited by child processes, at least not by default.  A few places in Twisted take care to set FD_CLOEXEC when they create a file descriptor, but they all have to duplicate the multiple fcntl calls necessary to do this and the logic to skip doing this is there is no FD_CLOEXEC constant available.  Lots of Twisted don't bother to do this (eg, UDP ports) and should be changed to do it.

Also, it's possible to make a handle non-inheritable on Windows, which is basically the same thing as this.  The appropriate platform calls should be made by a single implementation of this feature.

<details><summary>Searchable metadata</summary>

```
trac-id__3576 3576
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_cloexec_3576 branches/cloexec-3576
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1229123567000000 1229123567000000
changetime__1232129022000000 1232129022000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25651]) Branching to 'cloexec-3576'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I grabbed the code out of the [#10513](https://github.com/twisted/twisted/issues/10513) branch which was doing this.  It wasn't really central to the point of that changeset anymore.

I'm not really sure we need the Windows implementation included in the branch now.  Handles aren't inherited on Windows by default, and the process codepaths where the unsetting (ie, make inheritable) calls were happening won't ever run on Windows, since it has its own process implementation.

The code in the branch needs unit tests, which will be annoying, but probably pretty straightforward.  I don't think we should keep the Windows implementations unless we actually need them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

The description is a bit wrong.  I said that lots of Twisted doesn't bother to set CLOEXEC, but actually all the parts that I thought didn't are inheriting some fd creation behavior from someplace that does, so they're actually fine.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_5"></a>|@zseil commented|
|-|-|

There are still two places that don't set `FD_CLOEXEC`:
 * in `twisted.internet.posixbase`, the pipes or sockets created by the `_Waker.__init__()`,
 * in `twisted.internet.tcp`, the sockets accepted in `Port.doRead()`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Thanks, added close-on-exec to those places in r25827

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Okay

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @exarkun|
|-|-|

Good enough to merge, as far as I'm concerned.  Two suggestions, though:

  1. There's no particular reason to keep this new API private.  `setNonBlocking` hasn't given us any maintenance issues, has it?
  1. Conditionally defining functions is sucky.  I'd rather have a decorator like `doNothingOnWindows`.

Respond and merge at your leisure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26037]) Handle os.wait failing with EINTR

refs [#3576](https://github.com/twisted/twisted/issues/3576)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

  1. We can make them public later if there's demand for it.  Easier to do that than to do the opposite.  They're probably obscure enough that no one will care.  I think the existing functions haven't been a maintenance burden because no one cares about them, either. ;)  Note that unlike these new functions, the blocking functions are defined on all platforms and just blow up on Windows.
  1. If the Windows versions weren't no-ops _or_ if these were tests for functionality which wasn't so platform specific, I probably wouldn't have defined the functions as I did, but instead defined both sets separately and then bound names to one or the other depending on the platform.  Since the Windows versions are no-ops, there's hardly any reason to try to write tests for them, I think (the non-unit tests which exercise them make sure that they don't raise exceptions well enough for me, I don't think we need unit tests that directly call them and make sure they don't raise.  If they ever get expanded, I do think the code should be arranged slightly differently.).

The new fdesc tests failed on OS X because os.wait can fail with EINTR on OS X.  I fixed those (as seen above), so I'm merging now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [26045]) Copyright date updates

refs [#3576](https://github.com/twisted/twisted/issues/3576)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [26047]) use waitpid instead of wait to avoid getting nailed by a rogue child process

refs [#3576](https://github.com/twisted/twisted/issues/3576)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Actually, there was one further issue, fixed in r26047, which seems to have been caused by there being multiple processes ready to exit by the time the new fdesc tests ran.  This was only observed to fail on Python 2.3, not sure why.  The fix was to use waitpid to make sure to wait for the right process to exit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [26050]) Merge cloexec-3576

Author: glyph, exarkun, itamar
Reviewer: zseil, glyph
Fixes: [#3576](https://github.com/twisted/twisted/issues/3576)

Add (private) close-on-exec togglers to `twisted.internet.fdesc` and change
the rest of Twisted to use them instead of redoing the fcntl stuff directly
each time.  Also add close-on-exec to several file descriptor sources which
previously left that flag unset - the reactor waker's pipes and accepted
SOCK_STREAM sockets.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
